url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/ScriptyPanel.java

```diff
@@ -0,0 +1,205 @@
+{
+    private static class PanelEngine 
+    extends ScriptyCapable { }
+
+    private PanelEngine engine = new PanelEngine();
+    private JSplitPane splitter;
+    private boolean horizontalLayout = false;
+
+    public ScriptyPanel()
+    {
+        this.setLayout(new BorderLayout());
+        splitter = new JSplitPane(horizontalLayout?JSplitPane.HORIZONTAL_SPLIT:JSplitPane.VERTICAL_SPLIT);
+        buildPanel();
+    }
+    
+    public ScriptyPanel(ScriptyCapable aFacade)
+    {
+        this();
+        engine.setReplEngine(aFacade.getReplEngine());
+    }
+
+    public boolean isHorizontalLayout() 
+    {
+        return horizontalLayout;
+    }
+
+    public void setHorizontalLayout() 
+    {
+        this.horizontalLayout = true;
+        if(splitter != null) splitter.setOrientation(JSplitPane.HORIZONTAL_SPLIT);
+    }
+
+    public boolean isVerticalLayout()
+    {
+        return !horizontalLayout;
+    }
+
+    public void setVerticalLayout()
+    {
+        this.horizontalLayout = false;
+        if(splitter != null) splitter.setOrientation(JSplitPane.VERTICAL_SPLIT);
+    }
+
+    private void buildPanel()
+    {
+        final JToolBar lToolBar = new JToolBar();
+        this.add(lToolBar, BorderLayout.NORTH);
+
+        final JTextPane lOutputPane = new JTextPane();
+        lOutputPane.setEditable(false);
+        final JScrollPane lOutScroller = new JScrollPane(lOutputPane);
+        lOutScroller.setBackground(Color.white);
+        lOutScroller.setBorder(new TitledBorder("Output"));
+        final TextPaneWriter lWriter = new TextPaneWriter(lOutputPane);
+
+        final JTextPane lInputPane = new JTextPane();
+        final JScrollPane lInScroller = new JScrollPane(lInputPane);
+        lInScroller.setBackground(Color.white);
+        lInScroller.setBorder(new TitledBorder("Command Panel"));
+
+        splitter.add(lInScroller, JSplitPane.TOP);
+        splitter.add(lOutScroller, JSplitPane.BOTTOM);
+        this.add(splitter);
+
+        final SimpleAttributeSet lErrAttrs = new SimpleAttributeSet();
+        lErrAttrs.addAttribute(StyleConstants.CharacterConstants.Bold, Boolean.TRUE);
+        lErrAttrs.addAttribute(StyleConstants.CharacterConstants.Italic, Boolean.TRUE);
+        StyleConstants.setForeground(lErrAttrs, Color.RED);
+
+        final Action lExecuteAction = new AbstractAction("Run")
+        {
+            public void actionPerformed(ActionEvent e)
+            {
+                // Find top level frame.
+                final JFrame lFrame = (JFrame)SwingUtilities.getAncestorOfClass(JFrame.class, ScriptyPanel.this);
+
+                try
+                {
+                    // Set wait cursor.
+                    lFrame.setCursor(Cursor.getPredefinedCursor(Cursor.WAIT_CURSOR));
+                    //
+                    final String lCmd = lInputPane.getText();
+                    engine.getReplEngine().startNonInteractive(new StringReader(lCmd), new PrintWriter(lWriter), new PrintWriter(lWriter));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor: *PrintWriter*. Currently, there are execution paths that do not contain closure statements, for example, when ReplEngine.startNonInteractive throws an exception. Ensure the object returned by *PrintWriter()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).